python/ray/data/_internal/split.py

```diff
@@ -0,0 +1,438 @@
+    for b, m in blocks_with_metadata:
+        if m.num_rows is None:
+            num_rows = ray.get(get_num_rows.remote(b))
+        else:
+            num_rows = m.num_rows
+        if count >= index:
+            if not return_right_half:
+                break
+            right_blocks.append(b)
+            right_metadata.append(m)
+        elif count + num_rows <= index:
+            left_blocks.append(b)
+            left_metadata.append(m)
+        else:
+            b0, m0, b1, m1 = split_block_at_index.remote(
+                b, m, index - count, return_right_half
+            )
+            left_blocks.append(b0)
+            left_metadata.append(ray.get(m0))
+            right_blocks.append(b1)
+            right_metadata.append(ray.get(m1))
+        count += num_rows
+    return left_blocks, left_metadata, right_blocks, right_metadata
+
+
+def _split_at_indices(
+    blocks_with_metadata: Iterable[Tuple[ObjectRef[Block], BlockMetadata]],
+    indices: List[int],
+) -> Tuple[List[List[ObjectRef[Block]]], List[List[BlockMetadata]]]:
+    """Split blocks at the provided indices.
+
+    Args:
+        blocks_with_metadata: Block futures to split, including the associated metadata.
+        indices: The (global) indices at which to split the blocks.
+    Returns:
+        The block split futures and their metadata. If an index split is empty, the
+        corresponding block split future will be None.
+    """
+    # Implementation note: This splitting algorithm attempts to minimize the number of
+    # split tasks submitted and the amount of data moved/copied, such that:
+    #  - blocks that fit entirely within a split will be untouched,
+    #  - a block that constitutes multiple splits will be split via a single task with
+    #  multiple split returns,
+    #  - repeated splitting of the same block (via recursive splitting of leftovers)
+    #  should be minimal (a few cases remain but should only arise with very skewed
+    #  splits; see TODOs below),
+    #  - each split task will involve at most 1 block, for the sake of memory stability.
+
+    # The end-result is the number of split tasks should be
+    # O(min(num_splits, num_blocks)); explicitly,
+    #  - For split sizes >> block sizes (common case), split tasks will only be launched
+    #  and data will only be moved/copied for blocks crossing split boundaries, yielding
+    #  minimal number of tasks and amount of data movement.
+    #  - For split sizes << block sizes, 1-2 split tasks per block can be expected.
+    get_num_rows = cached_remote_fn(_get_num_rows)
+    split_block_at_index = cached_remote_fn(_split_block_at_index, num_returns=4)
+    split_block_at_indices = cached_remote_fn(_split_block_at_indices)
+
+    # Add infinite boundary index, which simplifies the splitting algorithm below.
+    indices = indices + [float("inf")]
+
+    # Queue of blocks that still need to be mapped to a split.
+    blocks = collections.deque(blocks_with_metadata)
```

Comment:
I am wondering if the ordering of original blocks matter or not. If we don't care about the ordering, maybe we can sort the blocks by number-of-rows, so it can minimize the copy/slice further.

E.g., blocks:
[b1, b2, b3]
b1 has 3 rows, b2 has 2 rows, and b3 has 2 rows. i.e. number of rows per block [3, 2, 2]
If each split size is 2, then we need to split/copy each block.

But if we sort the blocks to [b2, b3, b1]. i.e. number of rows per block [2, 2, 3].
We only need to split the last block - b1.

Reply:
Yeah, splitting is an internal operation and should not reorder rows. But we do reorder them when it's equal splitting which I think should be fixed.

Reply:
Yep ordering has to be preserved in `split_at_indices()`, but we essentially do what's described in `split()`, which does not make an order guarantee.